### PR TITLE
Better errors for types with same name

### PR DESCRIPTION
**Pull request by chrisbubernak**
Fix for issue #1419. Added a pretty unobtrusive fix but I'm wondering if I need to have more complicated logic (like typeToString() does) to handle certain cases that simply calling getFullyQualifiedName doesn't take care of.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1096,9 +1096,11 @@ module ts {
                 }
 
                 // Get qualified name 
-                if (enclosingDeclaration &&
+                if ((enclosingDeclaration &&
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: please find a way to simplify this 'if' conditional.  One option is to introduce locals above to make it clearer what you are testing for.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1237,10 +1239,10 @@ module ts {
                     }
                 }
 
-                function writeTypeofSymbol(type: ObjectType) {
+                function writeTypeofSymbol(type: ObjectType, flags?: TypeFormatFlags) {
                     writeKeyword(writer, SyntaxKind.TypeOfKeyword);
                     writeSpace(writer);
-                    buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Value);
+                    buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Value, undefined, flags);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: use a /*paramName:*/ comment to make it clear what 'undefined' means here.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -3508,7 +3510,13 @@ module ts {
                 }
                 if (reportErrors) {
                     headMessage = headMessage || Diagnostics.Type_0_is_not_assignable_to_type_1;
-                    reportError(headMessage, typeToString(source), typeToString(target));
+                    var sourceType = typeToString(source);
+                    var targetType = typeToString(target);
+                    if (sourceType !== targetType) {
+                        reportError(headMessage, sourceType, targetType);
+                    } else {
+                        reportError(headMessage, typeToString(source, undefined, TypeFormatFlags.UseFullyQualifiedType), typeToString(target, undefined, TypeFormatFlags.UseFullyQualifiedType));
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: use a /*paramName:*/ comment to make it clear what 'undefined' means here.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -3508,7 +3510,13 @@ module ts {
                 }
                 if (reportErrors) {
                     headMessage = headMessage || Diagnostics.Type_0_is_not_assignable_to_type_1;
-                    reportError(headMessage, typeToString(source), typeToString(target));
+                    var sourceType = typeToString(source);
+                    var targetType = typeToString(target);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: i would write this as:

```typescript
if (sourceType === targetType) {
    sourceType = typeToString(...);
    targetType = typeToString(...);
}

reportError(headMessage, sourceType, targetType)
```

File: src/compiler/checker.ts (hunk 2 above)
Comment (chrisbubernak) [inline src/compiler/checker.ts]: now that I'm passing a TypeFormatFlags parameter in here as well, does it make sense to change the parameter name from "flags" to "symbolFormatFlags" to make it less confusing?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -1095,11 +1095,11 @@ module ts {
...
                 }
 
-                // Get qualified name 
-                if (enclosingDeclaration &&
-                    // TypeParameters do not need qualification
-                    !(symbol.flags & SymbolFlags.TypeParameter)) {
-
+                // Get qualified name if there is an enclosing declaration and it is not a
+                // type parameter or we have a flag telling us to
+                var needFullyQualifiedName = (enclosingDeclaration && !(symbol.flags & SymbolFlags.TypeParameter));
```
Comment (sheetalkamat) [inline src/compiler/checker.ts]: Type parameters still shouldn't be qualified even when TypeFormatFlags.UseFullyQualifiedType

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -1193,17 +1193,17 @@ module ts {
...
                     else if (shouldWriteTypeOfFunctionSymbol()) {
-                        writeTypeofSymbol(type);
+                        writeTypeofSymbol(type, flags);
                     }
                     else if (typeStack && contains(typeStack, type)) {
                         // If type is an anonymous type literal in a type alias declaration, use type alias name
                         var typeAlias = getTypeAliasForTypeLiteral(type);
                         if (typeAlias) {
-                            buildSymbolDisplay(typeAlias, writer, enclosingDeclaration, SymbolFlags.Type);
+                            buildSymbolDisplay(typeAlias, writer, enclosingDeclaration, SymbolFlags.Type, /*flags*/ undefined, flags);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use `None`, not `undefined`.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -1122,7 +1122,7 @@ module ts {
                         writeTypeReference(<TypeReference>type, flags);
                     }
                     else if (type.flags & (TypeFlags.Class | TypeFlags.Interface | TypeFlags.Enum | TypeFlags.TypeParameter)) {
-                        buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Type);
+                        buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Type, /*flags*/ undefined, flags);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use `None`, not `undefined`.

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -1237,10 +1237,10 @@ module ts {
                     }
                 }
 
-                function writeTypeofSymbol(type: ObjectType) {
+                function writeTypeofSymbol(type: ObjectType, flags?: TypeFormatFlags) {
                     writeKeyword(writer, SyntaxKind.TypeOfKeyword);
                     writeSpace(writer);
-                    buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Value);
+                    buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Value, /*flags*/ undefined, flags);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use `None`, not `undefined`.

**Comment by DanielRosenwasser** [comment]
Need to bring this back up on the radar - will try to CR it when I get in.

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -1095,11 +1095,12 @@ module ts {
...
                 }
 
-                // Get qualified name 
-                if (enclosingDeclaration &&
-                    // TypeParameters do not need qualification
-                    !(symbol.flags & SymbolFlags.TypeParameter)) {
-
+                // Get qualified name if the symbol is not a type parameter
+                // and there is an enclosing declaration or we specifically
+				// asked for it
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Align indentation

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -1095,11 +1095,12 @@ module ts {
...
+                var typeParameter = (symbol.flags & SymbolFlags.TypeParameter);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Remove parens

File: src/compiler/checker.ts [hunk 11]
```diff
@@ -1122,7 +1123,7 @@ module ts {
                         writeTypeReference(<TypeReference>type, flags);
                     }
                     else if (type.flags & (TypeFlags.Class | TypeFlags.Interface | TypeFlags.Enum | TypeFlags.TypeParameter)) {
-                        buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Type);
+                        buildSymbolDisplay(type.symbol, writer, enclosingDeclaration, SymbolFlags.Type, SymbolFormatFlags.None, flags);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This is subtle - leave a comment saying that in these cases, the specified symbol flags need to be reinterpreted as type flags.

File: src/compiler/checker.ts [hunk 12]
```diff
@@ -1193,17 +1194,17 @@ module ts {
...
                     else if (shouldWriteTypeOfFunctionSymbol()) {
-                        writeTypeofSymbol(type);
+                        writeTypeofSymbol(type, flags);
                     }
                     else if (typeStack && contains(typeStack, type)) {
                         // If type is an anonymous type literal in a type alias declaration, use type alias name
                         var typeAlias = getTypeAliasForTypeLiteral(type);
                         if (typeAlias) {
-                            buildSymbolDisplay(typeAlias, writer, enclosingDeclaration, SymbolFlags.Type);
+                            buildSymbolDisplay(typeAlias, writer, enclosingDeclaration, SymbolFlags.Type, SymbolFormatFlags.None, flags);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Same comment as for `TypeFlags.Class | TypeFlags.Interface | TypeFlags.Enum | TypeFlags.TypeParameter`

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -1237,10 +1238,10 @@ module ts {
                     }
                 }
 
-                function writeTypeofSymbol(type: ObjectType) {
+                function writeTypeofSymbol(type: ObjectType, flags?: TypeFormatFlags) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Rename this to `typeFormatFlags`

**Comment by DanielRosenwasser** [comment]
Looking good so far! Just need to address the comments I've left.

**Comment by chrisbubernak** [comment]
Sweet, I'll get it fixed up when I get a chance.

**Comment by DanielRosenwasser** [comment]
LGTM :+1:

**Comment by chrisbubernak** [comment]
@DanielRosenwasser  Any other work that needs to be done here?

**Comment by DanielRosenwasser** [comment]
I don't think so, but I'd like to just get sign off from another member before pulling in the change.

@mhegazy @JsonFreeman @yuit

**Comment by DanielRosenwasser** [comment]
Thanks! This is going to save users a lot of mental anguish. :smile:

**Pull request merged**